### PR TITLE
Feature/save and fetch topics by section

### PR DESCRIPTION
**Pull request by Shawn-James**
# [User Can View Topics](HTTPS://TRELLO.COM) 

## Changes
- Reworked debugging, decoding, saving, and reading Topics
- Implemented sections using FRC
- Sort topics - newest first

## Checklist
`choose one minimum`
- [ ] created unit test
- [ ] updated unit test
- [ ] created UI test
- [ ] updated UI test
- [ ] bug fix/other

`required`
- [ ] tested locally
- [ ] updated the docs (https://github.com/Lambda-School-Labs/Labs26-Apollo-iOS-TeamA/blob/master/.github/documentation_standards.md)

`optional`
- [ ] added new dependencies

### CODE REVIEW

File: labs-ios-starter/Application/Network/NetworkService.swift [hunk 1]
```diff
@@ -16,7 +16,7 @@ typealias URLCompletion = ((Result<Data, ErrorHandler.NetworkError>) -> Void)
 /// Completion Handler with String .success and NetworkError .failure
 typealias CompleteWithString = (Result<String, ErrorHandler.NetworkError>) -> Void
 /// Completion Handler with [Topic] .success and NetworkError .failure
-typealias CompleteWithTopics = (Result<[Topic], ErrorHandler.NetworkError>) -> Void
+typealias CompleteWithTopics = (Result<[Topic]?, ErrorHandler.NetworkError>) -> Void
```
Comment (froggomad) [inline labs-ios-starter/Application/Network/NetworkService.swift]: The idea of the Result type is to guarantee a value on success and error on failure. If you need to unwrap something, that's ok - if the unwrapping fails, you can complete with error and return from the method
```suggestion
typealias CompleteWithTopics = (Result<[Topic], ErrorHandler.NetworkError>) -> Void
```

File: labs-ios-starter/Model/CoreData MO/Apollo/Topic/Topic/Topic+CoreDataClass.swift [hunk 2]
```diff
@@ -16,6 +16,7 @@ public final class Topic: NSManagedObject, Codable {
         case contextId = "contextid"
         case responsesToSend = "responses"
         case questionsToSend = "questions"
+        case timeStamp = "updated_at"
```
Comment (froggomad) [inline labs-ios-starter/Model/CoreData MO/Apollo/Topic/Topic/Topic+CoreDataClass.swift:19]: Awesome, it'll be great to be able to sort by most recent

File: labs-ios-starter/Surveys/Model Controllers/TopicController.swift [hunk 3]
```diff
@@ -70,36 +69,20 @@ class TopicController {
...
-                        if topic.leaderId == user.id { // member is leader
-                            userTopics.append(topic)
-                        } else if let members = topic.members { // member is user
-                            if members.contains(user) {
-                                userTopics.append(topic)
-                            }
-                        }
-                    }
-                }
-                completion(.success(userTopics))
```
Comment (froggomad) [inline labs-ios-starter/Surveys/Model Controllers/TopicController.swift:133]: Looks like you cleaned up a ton by adding the FRC!

File: labs-ios-starter/Surveys/Model Controllers/TopicController.swift [hunk 4]
```diff
@@ -70,36 +69,20 @@ class TopicController {
...
+
+                try? CoreDataManager.shared.saveContext()
+
+                completion(.success(nil))
```
Comment (froggomad) [inline labs-ios-starter/Surveys/Model Controllers/TopicController.swift]: I see why you changed to optional topics now, you can use the Void type in its place, and do

```suggestion
                completion(.success(Void()))
```

File: labs-ios-starter/Application/Network/NetworkService.swift (hunk 1 above)
Comment (froggomad) [reply labs-ios-starter/Application/Network/NetworkService.swift]: After reviewing the change that prompted this, You can replace `[Topic]` with `Void` here. If the completion handler is being used elsewhere, let's create a new CompletionHandler, or I believe there's an existing one with Void as the success Type

File: labs-ios-starter/Surveys/Model Controllers/TopicController.swift [hunk 5]
```diff
@@ -70,36 +69,20 @@ class TopicController {
...
 
             case let .failure(error):
-                // bubble error to caller
-                completion(.failure(error))
+                completion(.failure(error)) // bubble error to caller
```
Comment (froggomad) [inline labs-ios-starter/Surveys/Model Controllers/TopicController.swift:120]: amazing code quality

File: labs-ios-starter/Surveys/Model Controllers/TopicController.swift [hunk 6]
```diff
@@ -184,26 +167,27 @@ class TopicController {
...
-        func deleteTopic(topic: Topic, completion: @escaping CompleteWithNetworkError) {
-            guard
-                let id = topic.id,
-                let deleteRequest = createRequest(pathFromBaseURL: "topic/\(id)", method: .delete)
-            else {
-                completion(.failure(.badRequest))
-                return
-            }
+    /// Deletes a topic from the server
+    /// - Warning: This method should `ONLY` be called by the leader of a topic for their own survey
```
Comment (froggomad) [inline labs-ios-starter/Surveys/Model Controllers/TopicController.swift]: It looks like you're checking this on line 174 if I'm not mistaken and this comment can be removed 👍

**Review by froggomad — CHANGES_REQUESTED** [review]
Everything looks great, I'd just like to see the completion handler issue resolved (either in code or conversation :smile:)

Also, isn't it amazing how adding a FRC actually looks to reduce the total number of lines of code required? I usually avoid them in early production for sake of speed, but maybe I just need to get more familiar with implementing them!

File: labs-ios-starter/Application/Network/NetworkService.swift (hunk 1 above)
Comment (Shawn-James) [reply labs-ios-starter/Application/Network/NetworkService.swift]: Good call

File: labs-ios-starterTests/CoreDataTests.swift [hunk 7]
```diff
@@ -6,25 +6,35 @@
...
         } catch {
             XCTFail("error saving CoreData: \(error)")
         }
+    }
+
+    override func tearDownWithError() throws {
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
+    }
+
+    func testCanSaveAndFetchTopic() {
```
Comment (froggomad) [inline labs-ios-starterTests/CoreDataTests.swift:37]: Thanks for fixing this!

**Review by froggomad — APPROVED** [review]
Awesome, thanks - this will help me get through implementing question detail that much faster

**Pull request merged**